File: Práctica 1/S1/src/s1/Cliente.java

package s1;
import java.util.Scanner;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Random;

public class Cliente {
   ArrayList<Bicicleta> bicicletas;
   ArrayList<Carrera> carreras;
   factoriaCarrerayBicicleta factoria;
   boolean factoriasCreadas, tipoFactoria; //false -> montaña, true -> carretera
   private Scanner in;
   int codigo;
   
   Cliente(){
    bicicletas = new ArrayList<>();
    in = new Scanner(System.in);
    factoriasCreadas = false;
    carreras = new ArrayList<>();
    codigo = 0;
   }
   

   int leeEntero (int max, String msg1, String msg2) {
      Boolean ok;
      String cadena;
      int numero = -1;
      do {
        System.out.print (msg1);
        cadena = in.nextLine();
        try {  
          numero = Integer.parseInt(cadena);
          ok = true;
        } catch (NumberFormatException e) { // No se ha introducido un entero
          System.out.println (msg2);
          ok = false;  
        }
        if (ok && (numero < 0 || numero >= max)) {
          System.out.println (msg2);
          ok = false;
        }
      } while (!ok);

      return numero;
    }

    int menu (String titulo, ArrayList<String> lista) {
      String tab = "  ";
      int opcion;
      System.out.println (titulo);
      for (int i = 0; i < lista.size(); i++) {
        System.out.println (tab+i+"-"+lista.get(i));
      }

      opcion = leeEntero(lista.size(),
                            "\n"+tab+"Elige una opción: ",
                            tab+"Valor erróneo");
      return opcion;
    }
    
    public ArrayList<Carrera> getCarreras(){
        return this.carreras;
    }
  
    public  void gestionar() {
        int opcion = -1;

        ArrayList<String> opciones = new ArrayList<> 
        (Arrays.asList("Terminar","Construir factoria montaña","Construir factoria carretera","Crear bicicleta montaña","Crear bicicleta carretera",
                "Crear carrera montaña", "Crear carrera carretera"));
        
       while(opcion!=0){
           if(!this.factoriasCreadas){
            opciones = new ArrayList<> 
            (Arrays.asList("Terminar","Construir factoria montaña","Construir factoria carretera"));
            opcion = this.menu("¿Qué tipo de factoria deseas construir?", opciones);
            
            switch(opcion){
                case 1:
                    factoria = (factoriaCarrerayBicicleta) new factoriaMontana();
                    this.tipoFactoria = false;
                    this.factoriasCreadas = true;
                    System.out.println("Factoría de montaña creada");
                break;
                case 2:
                    factoria = (factoriaCarrerayBicicleta) new factoriaCarretera();
                    this.tipoFactoria = true;
                    this.factoriasCreadas = true;
                    System.out.println("Factoría de carretera creada");
            }
           }
           else{
                if(this.factoriasCreadas && !this.tipoFactoria){
                 opciones = new ArrayList<> 
                 (Arrays.asList("Terminar","Generar Participantes","Crear carrera montaña"));
                 opcion = this.menu("¿Qué deseas hacer?", opciones);
                }
                else if(this.factoriasCreadas && this.tipoFactoria){
                    opciones = new ArrayList<> 
                    (Arrays.asList("Terminar","Generar Participantes","Crear carrera carretera"));
                    opcion = this.menu("¿Qué deseas hacer?", opciones);
                }
                switch(opcion){
                    case 1:
                        this.bicicletas = generarParticipantes();
                        System.out.println("Participantes apuntados");
                    break;
                    case 2:
                        if(this.bicicletas.size()>0){
                            this.carreras.add(this.factoria.crearCarrera(codigo));
                            this.carreras.get(codigo).añadirParticipantes(bicicletas);
                            System.out.println("Carrera creada");
                            codigo++;

                            this.factoriasCreadas = false;                              //Ya has creado una carrera, volvemos al principio del menu
                        }
                        else
                            System.out.println("Necesitas crear al menos una bicicleta");
                    break;
                }         
            } 
        }    
    }
    
    public ArrayList<Bicicleta> generarParticipantes(){
        ArrayList<Bicicleta> bicis = new ArrayList();
        Random rand = new Random();
        int nparticipantes = rand.nextInt(10) + 10;

        for(int i=1;i<=nparticipantes;i++){
            bicis.add(this.factoria.crearBicicleta(i));
        }
        return bicis;
    }
}

File: Práctica 1/S1R/lib/Cliente.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
module Ruby
  class Cliente
    
    attr_accessor :bicicletas,:carreras,:factoria    

    
    
    def initialize 
      @bicicletas=Array.new
      @carreras=Array.new
      @codigo=0
    end
    
    def lee_entero(max,msg1,msg2)
      ok = false
      begin
        print msg1
        cadena = gets.chomp
        begin
          if (cadena =~ /\A\d+\Z/)
            numero = cadena.to_i
            ok = true
          else
            raise IOError
          end
        rescue IOError
          puts msg2
        end
        if (ok)
          if (numero >= max)
            ok = false
          end
        end
      end while (!ok)

      return numero
    end
    
    def menu(titulo,lista)
      tab = "  "
      puts titulo
      index = 0
      lista.each { |l|
        puts tab+index.to_s+"-"+l.to_s
        index += 1
      }

      opcion = lee_entero(lista.length,
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo")
      return opcion
    end
    
    
    def gestionar
      @opcion = -1
      @factoriaCreada = false
      @tipo 
           
      while @opcion != 0
        opciones = ["Terminar", "Construir factoria montania", "Construir factoria carretera"]
        @opcion = menu("¿Qué tipo de factoria quieres crear?",opciones)
        if @factoriaCreada == false
          case @opcion
          when 1
            factoria = FactoriaMontana.new
            @factoriaCreada=true
            @tipo=false
            
          when 2 
            factoria = FactoriaCarretera.new
            @factoriaCreada=true
            @tipo=true 
          end
        else
          if(factoriaCreada && !tipo)
            opciones = ["Terminar", "Generar participantes", "Crear carrera montania"]
            @opcion = menu("¿Qué desea hacer?",opciones)
             
          else
            if(factoriaCreada && tipo)
            opciones = ["Terminar", "Generar participantes", "Crear carrera carretera"]
            @opcion = menu("¿Qué desea hacer?",opciones)
            end
          end
          case @opcion
            when 1
              @bicicletas = generarParticipantes()
              puts "Participantes apuntados"
            when 2
              if(@bicicletas.length>0)
                carrera=@factoria.crearCarrera()
                carrera.codigo=@codigo
                @carreras<<carrera
                @codigo=@codigo+1
                
                @factoriaCreada=false
              else
                puts "Necesitas crear al menos una bicicleta"
              end
            end
          end
        end  
    end 
    
    def generarParticipantes
      bicis=Array.new 
      numPart=rand(1..10)+10
      for i in numPart
        bici=@factoria.crearBicicleta()
        bici.id=bici.id+i
        bicis<<bici
      end
    end
    
  end  
end